### PR TITLE
Update password

### PR DESCRIPTION
**Pull request by Surendhar1997**
Added changes for client password policy and update password for user.

### CODE REVIEW

**Review by serialcoder13 — COMMENTED** [review]
There are more coming

File: security/src/main/java/com/fincity/security/dao/ClientPasswordPolicyDAO.java [hunk 1]
```diff
@@ -1,19 +1,55 @@
...
+
+	public Mono<ClientPasswordPolicy> getByClientId(ULong clientId) {
+
+		return Mono.from(this.dslContext.selectFrom(SECURITY_CLIENT_PASSWORD_POLICY)
+		        .where(SECURITY_CLIENT_PASSWORD_POLICY.CLIENT_ID.eq(clientId)))
+		        .map(e -> e.into(ClientPasswordPolicy.class));
+
+	}
+
+	public Mono<Set<String>> getPastPasswords(ULong userId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/dao/ClientPasswordPolicyDAO.java:36]: This should be part of user service.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java [hunk 2]
```diff
@@ -409,4 +414,114 @@ public Mono<Set<ULong>> getUserListFromClientIds(Set<ULong> clientList) {
 
 		return Mono.just(users);
 	}
+
+	public Mono<Boolean> checkPasswordEqual(ULong userId, String newPassword) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/dao/UserDAO.java:418]: Please check the authentication service to see how to equal the password, it is not direct equals.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java [hunk 3]
```diff
@@ -409,4 +414,114 @@ public Mono<Set<ULong>> getUserListFromClientIds(Set<ULong> clientList) {
...
+		                .where(SECURITY_USER.ID.eq(userId))
+		                .limit(1))
+		        .map(Record1::value1)
+		        .map(pass -> pass.equals(newPassword));
+
+	}
+
+	public Mono<Set<String>> getPastPasswords(ULong userId) {
+
+		return Flux.from(
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/dao/UserDAO.java:433]: This can't be a set of strings. You need a list with the password types, too. If it is hashed, then you need to check differently. Please refer to Authentication Service.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java [hunk 4]
```diff
@@ -409,4 +414,114 @@ public Mono<Set<ULong>> getUserListFromClientIds(Set<ULong> clientList) {
...
+		        this.dslContext.select(SECURITY_PAST_PASSWORDS.PASSWORD)
+		                .from(SECURITY_PAST_PASSWORDS)
+		                .where(SECURITY_PAST_PASSWORDS.USER_ID.eq(userId)))
+		        .map(Record1::value1)
+		        .collectList()
+		        .map(HashSet::new);
+	}
+
+	public Mono<Boolean> updatePassword(ULong userId, String newPassword) {
+
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/dao/UserDAO.java:444]: You need to encrypt the password and add it to the past passwords simultaneously.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java [hunk 5]
```diff
@@ -409,4 +414,114 @@ public Mono<Set<ULong>> getUserListFromClientIds(Set<ULong> clientList) {
...
+		DeleteQuery<SecurityUserTokenRecord> query = this.dslContext.deleteQuery(SECURITY_USER_TOKEN);
+		query.addConditions(SECURITY_USER_TOKEN.ID.eq(tokenId));
+
+		return Mono.from(query)
+		        .map(val -> val > 0);
+
+	}
+
+	public Mono<List<String>> fetchPasswordsById(ULong userId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/dao/UserDAO.java:475]: Rename to fetchPastPasswords. Also, limit the query with the value in the password policy.

File: security/src/main/java/com/fincity/security/service/UserService.java [hunk 6]
```diff
@@ -400,5 +415,212 @@ public Mono<Boolean> removingPermissionFromUser(ULong userId, ULong permissionId
...
+			        return Mono.just(true);
+
+		        }
+
+		).log();
+
+	}
+
+	// delete past password record based on oldest time stamp for selected userId
+	public Mono<Boolean> deletePastPassword(ULong userId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/UserService.java:545]: You should not have a delete past password at all.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java [hunk 7]
```diff
@@ -409,4 +414,114 @@ public Mono<Set<ULong>> getUserListFromClientIds(Set<ULong> clientList) {
...
+
+		return Flux.from(this.dslContext.select(SECURITY_PAST_PASSWORDS.PASSWORD)
+		        .from(SECURITY_PAST_PASSWORDS)
+		        .where(SECURITY_PAST_PASSWORDS.USER_ID.eq(userId))
+		        .orderBy(SECURITY_PAST_PASSWORDS.CREATED_AT.asc()))
+		        .map(Record1::value1)
+		        .collectList();
+	}
+
+	public Mono<Boolean> deletePastPasswordBasedOnUserId(ULong userId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/dao/UserDAO.java:485]: You should not delete past password.

File: security/src/main/java/com/fincity/security/service/UserService.java [hunk 8]
```diff
@@ -400,5 +415,212 @@ public Mono<Boolean> removingPermissionFromUser(ULong userId, ULong permissionId
...
+	// selected user id
+
+	public Mono<Boolean> addBasedOnHistoryCount(ULong reqUserId, String password, ULong lUserId,
+	        ClientPasswordPolicy passwordPolicy) {
+
+		return flatMapMono(
+
+		        () -> Mono.just(reqUserId),
+
+		        userId -> this.dao.fetchPasswordsCount(userId),
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/UserService.java:521]: Usually, you don't count the passwords. You must insert it into past passwords and then write a delete query based on the client password policy. It has to be executed separately by calling subscribe method.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	private SecurityMessageResourceService securityMessageResourceService;
+
+	private final Set<Character> specialCharacters = new HashSet<>(
+	        Arrays.asList('~', '`', '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '-', '+', '=', '{', '}', '[',
+	                ']', '|', '\\', '/', ':', ';', '\"', '\'', '<', '>', ',', '.', '?'));
+
+	@PreAuthorize("hasAuthority('Authorities.Client_Password_Policy_UPDATE')")
+	@Override
+	protected Mono<ClientPasswordPolicy> updatableEntity(ClientPasswordPolicy entity) {
+		return super.update(entity);
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:40]: You have to restrict the client_id and other audit fields like created at, updated by, and so on.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	@Override
+	protected Mono<Map<String, Object>> updatableFields(ULong key, Map<String, Object> fields) {
+		return Mono.just(fields);
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:45]: You must restrict the client_id and other audit fields like created at, updated by, and so on.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	public Mono<Boolean> checkPasswordPolicyExists(ULong clientId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:48]: Why is this required?

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+		        (passwordPolicy, isAlphaNumberic, isSpecial) ->
+
+				!passwordPolicy.isSpacesAllowed() && password.contains(" ") ?
+
+				        Mono.just(false) : Mono.just(isAlphaNumberic && isSpecial),
+
+		        (passwordPolicy, isAlphaNumberic, isSpecial, isSpace) ->
+				{
+			        if (!isSpace.booleanValue())
+				        return securityMessageResourceService.throwMessage(HttpStatus.PRECONDITION_FAILED,
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:79]: Pre-condition failure is for a different reason. Please check the RFC. Use Bad Request HTTP status.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+		        },
+
+		        (passwordPolicy, isAlphaNumberic, isSpecial, isSpace, isRegex, isLength) ->
+
+				isLength.booleanValue() ? Mono.just(isAlphaNumberic && isSpecial && isSpace && isRegex) : Mono.empty(),
+
+		        (passwordPolicy, isAlphaNumberic, isSpecial, isSpace, isRegex, isLength, isValid) -> Mono
+		                .just(isAlphaNumberic && isSpecial && isSpace && isRegex && isValid)
+
+		// Add past passwords and history check later in that past passwords service
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:104]: Where are we doing the past passwords check?

File: security/src/main/java/com/fincity/security/service/UserService.java [hunk 14]
```diff
@@ -400,5 +415,212 @@ public Mono<Boolean> removingPermissionFromUser(ULong userId, ULong permissionId
...
+		).log()
+
+		        .switchIfEmpty(securityMessageResourceService.throwMessage(HttpStatus.FORBIDDEN,
+		                "Password cannot be updated"));
+
+	}
+
+	// add past password check here with dao and service here
+
+	public Mono<UShort> getHistoryCount(ULong clientId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/UserService.java:493]: Where are we using this?

File: security/src/main/java/com/fincity/security/service/UserService.java [hunk 15]
```diff
@@ -400,5 +415,212 @@ public Mono<Boolean> removingPermissionFromUser(ULong userId, ULong permissionId
...
+
+		        passwordPolicy -> passwordPolicy != null && passwordPolicy.getPassHistoryCount() != null
+		                ? Mono.just(passwordPolicy.getPassHistoryCount())
+		                : Mono.just(UShort.valueOf(0)));
+
+	}
+
+	// delete past password
+
+	public Mono<List<String>> fetchList(ULong userId) {
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/UserService.java:507]: Please remove methods if we are not using.

File: security/src/main/resources/messages_en.properties [hunk 16]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
 remove_permission_from_role_error=Unable to remove $ permission for the $ role
 remove_package_error=Unable to remove $ package from $ client
 role_remove_from_package_error=Unable to remove $ role from $ package
+client_password_policy_error=Given password is violating the policy which is defined by the client
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:25]: Password is violating the client's policy

File: security/src/main/resources/messages_en.properties [hunk 17]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+capital_letters_missing=Capital letters are missing in new password which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:26]: At least one capital case letter is required.

File: security/src/main/resources/messages_en.properties [hunk 18]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+small_letters_missing=Small letters are missing in new password which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:27]: At least one small case letter is required.

File: security/src/main/resources/messages_en.properties [hunk 19]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+numbers_missing=Numbers are missing in new password which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:28]: At least one number is required

File: security/src/main/resources/messages_en.properties [hunk 20]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+special_characters_missing =Special characters are missing in new password which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:29]: At least one special character is required

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+		                : Mono.empty(),
+
+		        (passwordPolicy, isAlphaNumberic) -> passwordPolicy.isAtleastOneSpecialChar()
+		                ? checkInSpecialCharacters(password).map(val -> val && isAlphaNumberic)
+		                : Mono.just(isAlphaNumberic),
+
+		        // re check here
+		        (passwordPolicy, isAlphaNumberic, isSpecial) ->
+
+				!passwordPolicy.isSpacesAllowed() && password.contains(" ") ?
```
Comment (serialcoder13) [inline security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:72]: The logic should be if the password contains a space, then check if it is allowed or not.

File: security/src/main/resources/messages_en.properties [hunk 22]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+spaces_missing=Spaces characters are missing in new password which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:30]: It has to be space is not allowed message. "Spaces are not allowed"

File: security/src/main/resources/messages_en.properties [hunk 23]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+regex_mismatch=Given new password is mismatch with the pattern which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:31]: The password doesn't match the suggested pattern.

File: security/src/main/resources/messages_en.properties [hunk 24]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+min_length_error=Given new password length is less than the minimum which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:32]: Password should be at least $ characters long.

File: security/src/main/resources/messages_en.properties [hunk 25]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+max_length_error=Given new password length is more than the maximum which is defined as per the client password policy
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:33]: Password should not be more than $ characters long.

File: security/src/main/resources/messages_en.properties [hunk 26]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+user_not_active=The user which is being trying to access is inactive
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:34]: The user is not active.

File: security/src/main/resources/messages_en.properties [hunk 27]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+old_new_password_match=Given new password is matching with the existing old password.Please Check.
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:35]: The password matches the existing password.

File: security/src/main/resources/messages_en.properties [hunk 28]
```diff
@@ -22,3 +22,15 @@ assign_permission_error_for_role=Unable to assign $ permission for $ role
...
+new_password_missing=Please provide the new password as it is required to update the password.
```
Comment (serialcoder13) [inline security/src/main/resources/messages_en.properties:36]: A password is required to update the password.

**Review by serialcoder13 — CHANGES_REQUESTED** [review]
Many many changes are required.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java (hunk 11 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:48]: Removed it. As it is not being used.

File: security/src/main/resources/messages_en.properties (hunk 16 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:25]: updated

File: security/src/main/resources/messages_en.properties (hunk 17 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:26]: updated

File: security/src/main/resources/messages_en.properties (hunk 18 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:27]: updated

File: security/src/main/resources/messages_en.properties (hunk 19 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:28]: updated

File: security/src/main/resources/messages_en.properties (hunk 20 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:29]: updated

File: security/src/main/resources/messages_en.properties (hunk 22 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:30]: updated

File: security/src/main/resources/messages_en.properties (hunk 23 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:31]: updated

File: security/src/main/resources/messages_en.properties (hunk 26 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:34]: updated

File: security/src/main/resources/messages_en.properties (hunk 28 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:36]: updated

File: security/src/main/resources/messages_en.properties (hunk 27 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:35]: updated

File: security/src/main/resources/messages_en.properties (hunk 24 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:32]: updated

File: security/src/main/resources/messages_en.properties (hunk 25 above)
Comment (Surendhar1997) [reply security/src/main/resources/messages_en.properties:33]: updated.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java (hunk 12 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:79]: updated

File: security/src/main/java/com/fincity/security/dao/ClientPasswordPolicyDAO.java (hunk 1 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/dao/ClientPasswordPolicyDAO.java]: moved it.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java (hunk 10 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:45]: if audit trail fields are sent in map removing it if present.

File: security/src/main/java/com/fincity/security/service/UserService.java (hunk 15 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/UserService.java]: removed

File: security/src/main/java/com/fincity/security/service/UserService.java (hunk 14 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/UserService.java:492]: removing this method as it not used.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java (hunk 21 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java]: updated need to test.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java (hunk 2 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/dao/UserDAO.java:418]: updated need to test

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java (hunk 9 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:41]: updated.

File: security/src/main/java/com/fincity/security/service/UserService.java (hunk 6 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/UserService.java:539]: removed this method

File: security/src/main/java/com/fincity/security/dao/UserDAO.java (hunk 7 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/dao/UserDAO.java:475]: removed this method

File: security/src/main/java/com/fincity/security/service/UserService.java (hunk 8 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/UserService.java:514]: removed this method from add past password method.

File: security/src/main/java/com/fincity/security/dao/UserDAO.java (hunk 3 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/dao/UserDAO.java:422]: updated

File: security/src/main/java/com/fincity/security/dao/UserDAO.java (hunk 5 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/dao/UserDAO.java]: not using this. Hence removed this dao method.

File: security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java (hunk 13 above)
Comment (Surendhar1997) [reply security/src/main/java/com/fincity/security/service/ClientPasswordPolicyService.java:110]: handled in user service when they only request for updating password.

**Review by serialcoder13 — APPROVED** [review]
